sql-streaming-mqtt/src/main/scala/org/apache/bahir/sql/streaming/mqtt/MQTTStreamSource.scala

```diff
@@ -0,0 +1,191 @@
+import scala.collection.mutable.ArrayBuffer
+import scala.util.{Failure, Success, Try}
+
+import org.apache.bahir.utils.Logging
+import org.eclipse.paho.client.mqttv3._
+import org.eclipse.paho.client.mqttv3.persist.{MemoryPersistence, MqttDefaultFilePersistence}
+
+import org.apache.spark.sql.{DataFrame, SQLContext}
+import org.apache.spark.sql.execution.streaming.{LongOffset, Offset, Source}
+import org.apache.spark.sql.sources.{DataSourceRegister, StreamSourceProvider}
+import org.apache.spark.sql.types.{StringType, StructField, StructType, TimestampType}
+
+
+object MQTTStreamConstants {
+
+  val DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
+
+  val SCHEMA_DEFAULT = StructType(StructField("value", StringType)
+    :: StructField("timestamp", TimestampType) :: Nil)
+}
+
+class MQTTTextStreamSource(brokerUrl: String, persistence: MqttClientPersistence,
+    topic: String, clientId: String, messageParser: Array[Byte] => (String, Timestamp),
+    sqlContext: SQLContext) extends Source with Logging {
+
+  override def schema: StructType = MQTTStreamConstants.SCHEMA_DEFAULT
+
+  private val store = new LocalMessageStore(persistence, sqlContext.sparkContext.getConf)
+
+  private val messages = new TrieMap[Int, (String, Timestamp)]
+
+  private val initLock = new CountDownLatch(1)
+
+  private var offset = 0
+
+  private var client: MqttClient = _
+
+  private def fetchLastProcessedOffset(): Int = {
+    Try(store.maxProcessedOffset) match {
+      case Success(x) =>
+        log.info(s"Recovering from last stored offset $x")
+        x
+      case Failure(e) => 0
+    }
+  }
+
+  initialize()
+  private def initialize(): Unit = {
+
+    client = new MqttClient(brokerUrl, clientId, persistence)
+    val mqttConnectOptions: MqttConnectOptions = new MqttConnectOptions()
+    mqttConnectOptions.setAutomaticReconnect(true)
+    // This is required to support recovery. TODO: configurable ?
+    mqttConnectOptions.setCleanSession(false)
+
+    val callback = new MqttCallbackExtended() {
+
+      override def messageArrived(topic_ : String, message: MqttMessage): Unit = synchronized {
+        initLock.await() // Wait for initialization to complete.
+        val temp = offset + 1
+        messages.put(temp, messageParser(message.getPayload))
+        offset = temp
+        log.trace(s"Message arrived, $topic_ $message")
+      }
+
+      override def deliveryComplete(token: IMqttDeliveryToken): Unit = {
+      }
+
+      override def connectionLost(cause: Throwable): Unit = {
+        log.warn("Connection to mqtt server lost.", cause)
+      }
+
+      override def connectComplete(reconnect: Boolean, serverURI: String): Unit = {
+        log.info(s"Connect complete $serverURI. Is it a reconnect?: $reconnect")
+      }
+    }
+    client.setCallback(callback)
+    client.connect(mqttConnectOptions)
+    client.subscribe(topic)
+    // It is not possible to initialize offset without `client.connect`
+    offset = fetchLastProcessedOffset()
```

Comment:
Hi Fred, offset is initialized here. Because it can only be fetched once we have connected the client. This also the reason, I am holding an `initlock`.